spec/services/display_ad_event_rollup_spec.rb

```diff
@@ -0,0 +1,39 @@
+require "rails_helper"
+
+RSpec.describe DisplayAdEventRollup, type: :service do
+  let(:ad1) { create :display_ad }
+  let(:ad2) { create :display_ad }
+  let(:user1) { create :user }
+
+  def override_timestamps
+    DisplayAdEvent.record_timestamps = false
+    yield
+    DisplayAdEvent.record_timestamps = true
+  end
+
+  def days_ago_as_range(num)
+    (Date.current - num.days).all_day
+  end
+
+  it "compacts display_ad_events" do
```

Comment:
This approach looks pretty great overall!

Since we didn't have much of a spec for this I want to be sure I can wrap my head around the grouping logic as to what gets rolled up, when and why.

I think adding some more specific tests to describe the different contexts in which events are rolled up into groups could ensure all scenarios are safely accounted for, and there is minimal gray area about how this is _supposed to_ work.

Does that make sense?

Reply:
:+1: Makes sense. This lost a lot of clarity when it moved away from SQL. `group by category, context_type, display_ad_id, user_id, created_on` isn't exactly _easy_, but it _easier_ than these nested data structures. These tests are relatively slow, so I didn't want to add too many if they weren't worth the cost, at the same time, I had a feeling this needed a few more.

Reply:
@jaw6 I _think_ the tests are worth it, even with the speed trade off, just to ensure nobody comes along and changes the functionality in ways that have systematic repercussions.

However, if you feel like the speed tradeoff isn't worth it, I think any comment or code naming adjustment that makes it abundantly clear how the rollup works is good enough.